### PR TITLE
Add events command to legal-hold

### PR DESCRIPTION
**Pull request by maddie-vargo**
This PR addresses Phase II of #176 to allow users to list legal hold administrative events. This command includes options to filter the results based on a specific legal hold uid, a beginning timestamp, an end timestamp, and a list of event types. 

**Testing Procedure**
```
code42 legal-hold events
code42 legal-hold --matter-id <matter id> --begin 30d --end 04/01/2021 --event-type MembershipCreated 
```

**PR Checklist**
Did you remember to do the below?

- [x]  Add unit tests to verify this change
- [x]  Add an entry to CHANGELOG.md describing this change
- [x]  Add docstrings for any new public parameters / methods / classes

### CODE REVIEW

File: CHANGELOG.md
Comment (unparalleled-js) [inline CHANGELOG.md:15]: Should the command be called `search-events` instead?

* It'd be a verb, like the rest of our commands
* It'd mirror the `file-events search` command

File: docs/userguides/legalhold.md
Comment (unparalleled-js) [inline docs/userguides/legalhold.md:102]: I think `time stamp` should be ` word, like `timestamp`

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:30]: We no longer need to use `OrderedDict` in the CLI, we no longer support any python version that requires that.
Regular `{}` syntax should suffice

Comment (unparalleled-js) [reply src/code42cli/cmds/legal_hold.py:30]: The above `_MATTER_KEYS_MAP`  could be changed as well

File: src/code42cli/cmds/legal_hold.py
Comment (timabrmsn) [inline src/code42cli/cmds/legal_hold.py:278]: Can we move the filtering logic out into the top-level events command function? That way the name of this function is more accurate.

So it would be like:

```
events = _get_all_events(state.sdk, matter_id, begin, end)
if event_type:
    events = [e for e in events if e["eventType"] == event_type]
if events:
    formatter.echo_formatted_list(events)
```

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:142]: Could you add the `-m` as a short flag for `matter_id_option`?

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:143]: Small refactor opportunity: we have matter_id_option already defined in this module.  We could create a shared method

```
def matter_id_option(required, help):
  return click.option(
    "-m",
    "--matter-id",
    required=required,
    type=str,
    help=help,
)
```

And use that for all matter_id options and then only have set two fields

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:158]: Missing `.` at end of help text

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:143]: Missing `.` at end of help text

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:160]: This should use the `begin_option` defined in `code42cli.options`
There will be several advantages to doing this.

Comment (unparalleled-js) [reply src/code42cli/cmds/legal_hold.py:160]: Same with `end_option`

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:165]: Missing period after help text

`Report on legal hold events.`

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:169]: `echo_formatted_list` works well with small lists, but we should handle larger lists via a pager.

Like this:

```
        if len(events) > 10:
            output = formatter.get_formatted_output(events)
            click.echo_via_pager(output)
        else:
            formatter.echo_formatted_list(events)
```

File: tests/cmds/test_legal_hold.py
Comment (unparalleled-js) [inline tests/cmds/test_legal_hold.py:260]: It is better to have mock generator methods actually be generators and not just lists.  I know it probably works fine, but I have seen bugs show up because the tests were doing this. There are some subtle differences.

File: tests/cmds/test_legal_hold.py
Comment (unparalleled-js) [inline tests/cmds/test_legal_hold.py:641]: From the test alone, it is not immediately clear why these values are expected.
Maybe a comment saying `// From `all_events_response` would help

File: tests/cmds/test_legal_hold.py
Comment (unparalleled-js) [inline tests/cmds/test_legal_hold.py:653]: Could we get test(s) for `begin_date` / `end_date` validation?
It could be to just assert the values were passed into py42 correctly.

File: src/code42cli/cmds/legal_hold.py
Comment (maddie-vargo) [reply src/code42cli/cmds/legal_hold.py:160]: Sure, I avoided this option initially because the `begin` option requires a cursor initialized for state, which I didn't think was necessary. I've added a LegalHoldEvents cursor to the cursor_store.py to make it work, but not sure if that's appropriate.

Comment (timabrmsn) [reply src/code42cli/cmds/legal_hold.py:160]: I wonder if we should implement the `--use-checkpoint` option for this command and probably also make a `send-to`. As I could see customers wanting to be able to automate sending this data into a SIEM.

Comment (unparalleled-js) [reply src/code42cli/cmds/legal_hold.py:160]: Oh!  @maddie-vargo Sorry, I did not forget about that when I made my initial comments.

I do like the idea from @timabrmsn, supporting checkingpointing would be a nice feature. We can always add it later though

File: CHANGELOG.md
Comment (maddie-vargo) [reply CHANGELOG.md]: Changed to `search-events` > let me know if this is not what you had in mind

File: tests/cmds/test_legal_hold.py
Comment (maddie-vargo) [reply tests/cmds/test_legal_hold.py:297]: Let me know if I'm still missing the mark here.

File: tests/cmds/test_legal_hold.py
Comment (maddie-vargo) [reply tests/cmds/test_legal_hold.py]: Defined variables at the top here. I don't love the names

```
OLDER_LEGAL_HOLD_CREATED_EVENT = "564564654566"
NEWER_LEGAL_HOLD_MEMBERSHIP_EVENT = "74533457745"
```

File: src/code42cli/cmds/legal_hold.py
Comment (maddie-vargo) [reply src/code42cli/cmds/legal_hold.py:158]: @unparalleled-js  would it be possible to use the `begin_option` but somehow designate it as optional? I left the original options in there for now. 

I left checkpoint'ing off for now.

File: tests/cmds/test_legal_hold.py
Comment (maddie-vargo) [reply tests/cmds/test_legal_hold.py:699]: I'll work on this. I had tried this initially, but had trouble getting a test to recognize the date inputs in the runner.

Comment (maddie-vargo) [reply tests/cmds/test_legal_hold.py:707]: I added a test titled `test_search_events_is_called_with_expected_begin_timestamp`

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:74]: @annie-payseur When you have chance, could you help review these option help texts? I will tag you in spots.

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:163]: @annie-payseur This is the command's help text: `Report on legal hold events.`

File: tests/cmds/test_legal_hold.py
Comment (unparalleled-js) [inline tests/cmds/test_legal_hold.py:176]: Something seems off about this response.... It is not valid JSON. Should these objects be under a list with key: `legalHoldEvents`

Comment (unparalleled-js) [reply tests/cmds/test_legal_hold.py:176]: Or else I believe line 290 will create a response where this a `str` type (I think).

File: tests/cmds/test_legal_hold.py
Comment (unparalleled-js) [inline tests/cmds/test_legal_hold.py:289]: Is this fixture used anywhere? I am not seeing.
Perhaps this and `EVENTS_RESPONSE` can be removed?

File: tests/cmds/test_legal_hold.py
Comment (unparalleled-js) [reply tests/cmds/test_legal_hold.py:176]: I don't think the fixture that uses this is ever used, if that is the, case can this be removed?

File: tests/cmds/test_legal_hold.py
Comment (unparalleled-js) [reply tests/cmds/test_legal_hold.py]: The names are fine..

but alternatively they could be called 
`_CREATE_EVENT_ID` and `_MEMBERSHIP_EVENT_ID`

Reasons:
1.) they are already in a legal hold module so don't need the prefix,
2.) they can be internal since nothing should be importing or using them outside of this test module, and 
3.) ending with ID makes it more apparent that it's not a JSON object

File: tests/cmds/test_legal_hold.py
Comment (unparalleled-js) [reply tests/cmds/test_legal_hold.py]: This works!
You could add more events to the list so that it yields more than once, that way it is more like what will happen in real life (unless there is often only a single event). Not a big deal though.

File: tests/cmds/test_legal_hold.py
Comment (maddie-vargo) [reply tests/cmds/test_legal_hold.py]: I took your suggestions. Thanks!

File: tests/cmds/test_legal_hold.py
Comment (maddie-vargo) [reply tests/cmds/test_legal_hold.py:176]: Yes, that is no longer needed, since I moved to a generator format for the mocker. I forgot to take it out on my last push along with it's corresponding fixture.

File: tests/cmds/test_legal_hold.py
Comment (maddie-vargo) [reply tests/cmds/test_legal_hold.py:289]: Yes, this is no longer needed. See my comment on the item above.

File: src/code42cli/cmds/legal_hold.py
Comment (annie-payseur) [reply src/code42cli/cmds/legal_hold.py:86]: Looks good, thanks!

**Review by unparalleled-js — APPROVED** [review]

File: src/code42cli/cmds/legal_hold.py
Comment (annie-payseur) [reply src/code42cli/cmds/legal_hold.py:163]: I suggest changing to `Tools for getting legal hold event data.` (which is more consistent with the help text for `security-data`)

**Comment by maddie-vargo** [comment]
Does anything else need to happen on this PR, or can it be merged and closed?

**Comment by unparalleled-js** [comment]
> Does anything else need to happen on this PR, or can it be merged and closed?

Was just waiting for a bugfix release to happen (just did). I can merge this now, but do you mind fixing the CHANGELOG one last time?  Just need to add your changed to a new Unreleased section (above the version section that was released today). After that, I will merge this in and get it officially tested and release it ASAP.

**Review by unparalleled-js — APPROVED** [review]

**Pull request merged**
